modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C

```diff
@@ -0,0 +1,170 @@
+    mooseError("Sampler number of columns does not match number of inputted distributions.");
+
+  _covar_function = CovarianceFunction::makeCovarianceKernel(_kernel_type, this);
+}
+
+void
+GaussianProcessTrainer::execute()
+{
+  // Check if results of samples matches number of samples
+  __attribute__((unused)) dof_id_type num_rows =
+      _values_distributed ? _sampler->getNumberOfLocalRows() : _sampler->getNumberOfRows();
+  mooseAssert(num_rows == _values_ptr->size(),
+              "Sampler number of rows does not match number of results from vector postprocessor.");
+
+  unsigned int _num_samples = _values_ptr->size();
+
+  // Offset for replicated/distributed result data
+  dof_id_type offset = _values_distributed ? _sampler->getLocalRowBegin() : 0;
+
+  // May load a very large matrix, which could be bad.
+  DenseMatrix<Real> params = _sampler->getLocalSamples();
+  _training_params.resize(params.m(), params.n());
+  for (unsigned int ii = 0; ii < _training_params.rows(); ii++)
+  {
+    for (unsigned int jj = 0; jj < _training_params.cols(); jj++)
+    {
+      _training_params(ii, jj) = params(ii, jj);
+    }
+  }
+
+  if (_standardize_params)
+  {
+    // comptue mean
+    _training_params_mean = _training_params.colwise().mean();
+    // center params
+    _training_params = _training_params.rowwise() - _training_params_mean.transpose();
+    // compute variance
+    _training_params_var = _training_params.colwise().squaredNorm() / _num_samples;
+    // scale params using std
+    _training_params =
+        _training_params.array().rowwise() / _training_params_var.transpose().array().sqrt();
+  }
+  else
+  {
+    // if not standardizing data set mean=0, std=1 for use in surrogate
+    _training_params_mean = RealEigenVector::Zero(_n_params);
+    _training_params_var = RealEigenVector::Ones(_n_params);
+  }
+
+  // Load training data into Eigen::Vector for easy manipulation
+  _training_data.resize(_num_samples, 1);
+  for (unsigned int ii = 0; ii < _num_samples; ii++)
+  {
+    _training_data(ii, 0) = (*_values_ptr)[ii - offset];
+  }
+
+  if (_standardize_data)
+  {
+    // comptue mean
+    _training_data_mean = _training_data.colwise().mean();
+    // center data
+    _training_data = _training_data.rowwise() - _training_data_mean.transpose();
+    // compute variance
+    _training_data_var = _training_data.colwise().squaredNorm() / _num_samples;
+    // scale data using std
+    _training_data =
+        _training_data.array().rowwise() / _training_data_var.transpose().array().sqrt();
+  }
+  else
+  {
+    // if not standardizing data set mean=0, std=1 for use in surrogate
+    _training_data_mean = RealEigenVector::Zero(1);
+    _training_data_var = RealEigenVector::Ones(1);
+  }
+
+  _K = _covar_function->compute_K(_training_params, _training_params, true);
+  _K_cho_decomp = _K.llt();
+  _K_results_solve = _K_cho_decomp.solve(_training_data);
+}
+
+void
+GaussianProcessTrainer::finalize()
+{
+}
```

Comment:
This is where the MPI communication should take place. This is probably why your implementation isn't working in parallel.

Reply:
I may need a bit of help on to how to properly use this

Reply:
It's "working". The GP is NOT able to take advantage of parallel run at the moment, but it works (in a serial manner) when run using multiple processes.